src/mobu/services/repo_cache.py

```diff
@@ -0,0 +1,114 @@
+
+from ..storage.git import Git
+
+__all__ = ["ClonedRepoInfo", "RepoCache"]
+
+
+@dataclass
+class ClonedRepoInfo:
+    """Information about a cloned git repo."""
+
+    dir: TemporaryDirectory
+    path: Path
+    hash: str
+
+
+class RepoCache:
+    """Coordinates cloning and filtering notebook repos.
+
+    Businesses can use this to avoid cloning the same repo multiple times per
+    flock.
+
+    Parameters
+    ----------
+    logger
+        A logger
+    """
+
+    def __init__(self, logger: BoundLogger) -> None:
+        self._dir = TemporaryDirectory(delete=False, prefix="mobu-notebooks-")
+        self._cache: dict[str, ClonedRepoInfo] = {}
+        self._add_lock = Lock()
+        self._remove_lock = Lock()
+        self._logger = logger
+
+    async def add(self, url: str, ref: str) -> ClonedRepoInfo:
+        """Clone a git repo and add it to the cache.
+
+        Parameters
+        ----------
+        url
+            The URL of the repo to clone
+        ref
+            The git ref to checkout after the repo is cloned
+        """
+        key = self._make_key(url=url, ref=ref)
+
+        async with self._add_lock:
+            # If the notebook repo has already been cloned, return the info
+            if info := self._cache.get(key):
+                self._logger.info(
+                    "Notebook repo cached", repo_url=url, ref=ref
+                )
+                return info
+
+            # If not, clone the repo
+            self._logger.info("Cloning notebook repo", repo_url=url, ref=ref)
+            repo_dir = TemporaryDirectory(delete=False, dir=self._dir.name)
+            with capturing_start_span(op="clone_repo"):
+                git = Git(logger=self._logger)
+                git.repo = Path(repo_dir.name)
+                await git.clone(url, repo_dir.name)
+                await git.checkout(ref)
+                repo_hash = await git.repo_hash()
+
+            info = ClonedRepoInfo(
+                dir=repo_dir, path=Path(repo_dir.name), hash=repo_hash
+            )
+
+            # Update the cache with the cloned repo's info
+            self._cache[key] = info
+            return info
+
+    async def remove(self, url: str, ref: str, repo_hash: str) -> None:
+        """Delete a cloned repo and remove it from the cache.
+
+        Only remove it from the cache if the hashes match. If they don't, that
+        means it was already removed and recloned.
+
+        Parameters
+        ----------
+        url
+            The URL of the repo to clone
+        ref
+            The git ref to checkout after the repo is cloned
+        hash
+            The hash of the cloned repo to remove
+        """
+        key = self._make_key(url=url, ref=ref)
+
+        # This theoretically doesn't need a lock, but if we add any awaits here
+        # in the future, we'd need to add a lock, and it would be pretty easy
+        # to forget to do it.
+        async with self._remove_lock:
```

Comment:
It feels weird to me to have two separate locks here instead of one protecting all manipulations of the underlying file system. I *think* this is safe, or at least I can't construct a scenario where it's a problem, but I think ti's more obvious if you use one lock, and I don't think it will matter much.